### PR TITLE
[WIP] [#138] import cards

### PR DESCRIPTION
**Pull request by mrzinger**
While I was working on implementation it become non-feature, but I managed to implement client side csv conversion and don't want to throw it away.
Please review and let me know if I can finalize it?
What is left
- [x] Fix jshint and stylehint exceptions
- [x] Fix import dialog layout
- [x] After you import a file, when you click again to import another file, the columns will be pre loaded from the previous file, maybe we could clean that up first.
- [x] After you submit a file, we are letting the users click on "import" even if there is no mapping added. You could disable the button and enable if there is 1 or more mappings added.
- [x] When there is an error on the file format (for example: duplicated headers), we are not displaying any error message and we are letting the user click on import even if there is no mapping.
- [x] If the file is empty, you are not displaying any error message as well.
- [x] ~~Add possibility to configure parsing options (?)~~ Papaparser automatically detects delimiter and other settings
- [ ] Refactoring (?)

### CODE REVIEW

**Comment by glauberramos** [comment]
Hey @mrzinger I will take a look on it as soon as I can, I am traveling right now so I have limited internet access. Thanks for you help!

**Comment by glauberramos** [comment]
Sorry for the change, but we changed the labels, the issues people should pick up are the  one with "priority". The ones that have "idea" label means we are not 100% sure that it needs to be implemented, they are just ideas for now, and we need or to detail more, or to make sure many users need that.

**Comment by mrzinger** [comment]
No problem @glauberramos. Looking forward to your feedback. 

P.S. At least me as a user would highly appreciate this functionality :)

**Comment by glauberramos** [comment]
@mrzinger can you post here a sample .csv?

**Comment by glauberramos** [comment]
It's fine, I don't need a sample .csv anymore, just created one here to test the feature.

**Comment by mrzinger** [comment]
[Documents.zip](https://github.com/funretro/distributed/files/943954/Documents.zip)
Tried with two formats: standard where the values are separated by "," and MS Office where ";" is used as separator. Papaparser parses both without any problems which is great.

**Comment by glauberramos** [comment]
I tested the feature and it is working really well, I was able to import a couple of files.

I just have some comments of things that I found that needs to be fixed:

1 - After you import a file, when you click again to import another file, the columns will be pre loaded from the previous file, maybe we could clean that up first.
2 - After you submit a file, we are letting the users click on "import" even if there is no mapping added. You could disable the button and enable if there is 1 or more mappings added.
2 - When there is an error on the file format (for example: duplicated headers), we are not displaying any error message and we are letting the user click on import even if there is no mapping.
3 - If the file is empty, you are not displaying any error message as well.

What do you think?

File: components/dialogs.html [hunk 1]
```diff
@@ -43,3 +43,22 @@
   <button ng-controller="MessageCtrl" class="delete-button" ng-click="dropped(dragEl, dropEl)">Merge</button>
   <a ng-click="closeAllModals()">Cancel</a>
 </script>
+<script type="text/ng-template" id="importCards">
+  <h1>Import cards</h1>
+  <span >
+    <p>1. <a style="margin:0" type="file" ngf-select="submitImportFile($file)" ngf-pattern="text/csv" accept="text/csv" ngf-max-size="10KB"> Submit</a> a csv file </p>
```
Comment (glauberramos) [inline components/dialogs.html]: We should not use inline style here. Please put it on a .scss file.

File: components/dialogs.html [hunk 2]
```diff
@@ -43,3 +43,22 @@
...
+    <p>2. Setup fields mapping</button></p>
+    <table>
+      <tr ng-repeat="column in import.mapping track by $index">
+        <td style="text-align:right">{{column.name}}</td>
```
Comment (glauberramos) [inline components/dialogs.html]: No inline style

File: components/dialogs.html [hunk 3]
```diff
@@ -43,3 +43,22 @@
...
+        <td style="text-align:left">
```
Comment (glauberramos) [inline components/dialogs.html]: No inline style

File: components/dialogs.html [hunk 4]
```diff
@@ -43,3 +43,22 @@
...
+          <select style="width:138" ng-model="column.mapFrom">
```
Comment (glauberramos) [inline components/dialogs.html]: No inline style

**Comment by mrzinger** [comment]
Thank you for the review! All the points sound reasonable. I will try to fix them as soon as possible.

File: test/mainControllerTest.js [hunk 5]
```diff
@@ -329,4 +329,91 @@ describe('MainCtrl: ', function() {
       expect(closeAllSpy.called).to.be.true;
     });
   });
+  
+  describe('Import', function() {
```
Comment (glauberramos) [inline test/mainControllerTest.js]: We should add tests for error as well, duplicate header, empty file, wrong .csv format file (like not using commas)

**Comment by glauberramos** [comment]
There is more information about errors here: http://papaparse.com/docs#errors

It says that it may contain errors but the parsing may work as well. Maybe we should only display a message if there is not data parsed (like empty file, all duplicate headers, totally wrong file format)

What do you think?

File: components/dialogs.html (hunk 1 above)
Comment (mrzinger) [inline components/dialogs.html]: Fixed

File: components/dialogs.html (hunk 2 above)
Comment (mrzinger) [inline components/dialogs.html]: Fixed

File: components/dialogs.html (hunk 3 above)
Comment (mrzinger) [inline components/dialogs.html]: Fixed

File: components/dialogs.html (hunk 4 above)
Comment (mrzinger) [inline components/dialogs.html]: Fixed

File: test/mainControllerTest.js (hunk 5 above)
Comment (mrzinger) [inline test/mainControllerTest.js]: Added test for empty file, I think that other kind of errors should be tested by Papaparser. I just simply show the error it generates

Comment (mrzinger) [inline test/mainControllerTest.js]: added simple test for malformed file import

**Comment by mrzinger** [comment]
Ready for review

**Comment by mrzinger** [comment]
Ready for 2nd round of review

**Comment by glauberramos** [comment]
I think everything is fine now. I will merge the PR but not deploy to production right now. I need to check some other changes and will deploy in the next days. Thanks for the work, it is working really well! 💃 💃 🕺 🎉

**Comment by mrzinger** [comment]
Awsome, thanks! I delete the branch
